compiler/rustc_query_system/src/query/plumbing.rs

```diff
@@ -398,48 +402,53 @@ where
     }
 }
 
+// Fast path for when incr. comp. is off.
 #[inline(always)]
-fn execute_job<Q, Qcx>(
+fn execute_job_non_incr<Q, Qcx>(
     query: Q,
     qcx: Qcx,
     key: Q::Key,
-    mut dep_node_opt: Option<DepNode<Qcx::DepKind>>,
     job_id: QueryJobId,
 ) -> (Q::Value, DepNodeIndex)
 where
     Q: QueryConfig<Qcx>,
     Qcx: QueryContext,
 {
-    let dep_graph = qcx.dep_context().dep_graph();
-    let dep_graph_data = match dep_graph.data() {
-        // Fast path for when incr. comp. is off.
-        None => {
-            // Fingerprint the key, just to assert that it doesn't
-            // have anything we don't consider hashable
-            if cfg!(debug_assertions) {
-                let _ = key.to_fingerprint(*qcx.dep_context());
-            }
+    // Fingerprint the key, just to assert that it doesn't
+    // have anything we don't consider hashable
+    if cfg!(debug_assertions) {
+        let _ = key.to_fingerprint(*qcx.dep_context());
+    }
 
-            let prof_timer = qcx.dep_context().profiler().query_provider();
-            let result =
-                qcx.start_query(job_id, query.depth_limit(), None, || query.compute(qcx, key));
-            let dep_node_index = dep_graph.next_virtual_depnode_index();
-            prof_timer.finish_with_query_invocation_id(dep_node_index.into());
-
-            // Similarly, fingerprint the result to assert that
-            // it doesn't have anything not considered hashable.
-            if cfg!(debug_assertions) && let Some(hash_result) = query.hash_result()
-            {
-                qcx.dep_context().with_stable_hashing_context(|mut hcx| {
-                    hash_result(&mut hcx, &result);
-                });
-            }
+    let prof_timer = qcx.dep_context().profiler().query_provider();
+    let result = qcx.start_query(job_id, query.depth_limit(), None, || query.compute(qcx, key));
+    let dep_node_index = qcx.dep_context().dep_graph().next_virtual_depnode_index();
```

Comment:
Could you encapsulate `next_virtual_depnode_index`? The goal is to make sure that it is only called when `DepGraph::data` is `None`.

For instance, making `DepGraph::Data` return some marker type `DisabledDepGraph` on which we can call that method?

Reply:
I tried to make a marker type `struct DisabledDepGraph<'a, K: DepKind>(&'a DepGraph<K>)`, but it doesn't seem like LLVM was able to optimize it away:
<table><tr><td rowspan="2">Benchmark</td><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th></tr><tr><td align="right">Time</td><td align="right">Time</td><td align="right">%</th></tr><tr><td>🟣 <b>clap</b>:check</td><td align="right">1.7102s</td><td align="right">1.7175s</td><td align="right"> 0.43%</td></tr><tr><td>🟣 <b>hyper</b>:check</td><td align="right">0.2533s</td><td align="right">0.2538s</td><td align="right"> 0.23%</td></tr><tr><td>🟣 <b>regex</b>:check</td><td align="right">0.9534s</td><td align="right">0.9567s</td><td align="right"> 0.35%</td></tr><tr><td>🟣 <b>syn</b>:check</td><td align="right">1.5869s</td><td align="right">1.5971s</td><td align="right"> 0.64%</td></tr><tr><td>🟣 <b>syntex_syntax</b>:check</td><td align="right">6.1097s</td><td align="right">6.1317s</td><td align="right"> 0.36%</td></tr><tr><td>Total</td><td align="right">10.6134s</td><td align="right">10.6568s</td><td align="right"> 0.41%</td></tr><tr><td>Summary</td><td align="right">1.0000s</td><td align="right">1.0040s</td><td align="right"> 0.40%</td></tr></table>

Reply:
What about `DisabledDepGraph(Lrc<AtomicU32>)`, and have the dep-graph hand out references to that struct?

Reply:
That sounds very performance equivalent.

Reply:
How about adding `assert!(self.data.is_none());` to `DepGraph::next_virtual_depnode_index()`?

Reply:
That seems to be cheaper:
<table><tr><td rowspan="2">Benchmark</td><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th></tr><tr><td align="right">Time</td><td align="right">Time</td><td align="right">%</th></tr><tr><td>🟣 <b>clap</b>:check</td><td align="right">1.7151s</td><td align="right">1.7189s</td><td align="right"> 0.22%</td></tr><tr><td>🟣 <b>hyper</b>:check</td><td align="right">0.2516s</td><td align="right">0.2525s</td><td align="right"> 0.34%</td></tr><tr><td>🟣 <b>regex</b>:check</td><td align="right">0.9532s</td><td align="right">0.9538s</td><td align="right"> 0.06%</td></tr><tr><td>🟣 <b>syn</b>:check</td><td align="right">1.5406s</td><td align="right">1.5400s</td><td align="right"> -0.04%</td></tr><tr><td>🟣 <b>syntex_syntax</b>:check</td><td align="right">5.9116s</td><td align="right">5.9193s</td><td align="right"> 0.13%</td></tr><tr><td>Total</td><td align="right">10.3722s</td><td align="right">10.3845s</td><td align="right"> 0.12%</td></tr><tr><td>Summary</td><td align="right">1.0000s</td><td align="right">1.0014s</td><td align="right"> 0.14%</td></tr></table>

There's a 0.01% code size increase so it probably doesn't get optimized away.